File: app/src/main/java/com/pacmac/devinfo/display/DisplayUtils.java
package com.pacmac.devinfo.display;

import android.content.Context;
import android.content.res.Configuration;
import android.graphics.Point;
import android.util.DisplayMetrics;
import android.view.Display;

import com.pacmac.devinfo.R;
import com.pacmac.devinfo.UIObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class DisplayUtils {

    public static final String EXPORT_FILE_NAME = "display_info";

    static UIObject getDensity(Context context, DisplayMetrics metrics) {
        return new UIObject(context.getString(R.string.display_density), String.format(Locale.ENGLISH, "%d", metrics.densityDpi), "dpi");
    }

    static UIObject getScaleFactor(Context context, DisplayMetrics metrics) {
        return new UIObject(context.getString(R.string.display_scale_factor), String.format(Locale.ENGLISH, "%.1f", metrics.density));
    }

    static UIObject getRefreshRate(Context context, Display display) {
        return new UIObject(context.getString(R.string.display_refresh_rate), String.format(Locale.ENGLISH, "%.1f", display.getRefreshRate()), "fps");
    }

    static List<UIObject> getResolution(Context context, Display display, DisplayMetrics metrics) {
        List<UIObject> list = new ArrayList<>();
        Point size = new Point();
//        int screenWidth = metrics.widthPixels;
//        int screenHeight = metrics.heightPixels;
        display.getRealSize(size);
        int screenWidth = size.x;
        int screenHeight = size.y;
        double widthInch = (double) screenWidth / (double) metrics.xdpi;
        double heightInch = (double) screenHeight / (double) metrics.ydpi;
        double widthDP = Math.pow(widthInch, 2);
        double heightDP = Math.pow(heightInch, 2);
        double diagonal = Math.sqrt(widthDP + heightDP);

        list.add(new UIObject(context.getString(R.string.display_resolution), String.format(Locale.ENGLISH, "%dx%d", screenWidth, screenHeight), "px"));
        list.add(new UIObject(context.getString(R.string.display_dimension), String.format(Locale.ENGLISH, "%.2fx%.2f", widthInch, heightInch), "in"));
        list.add(new UIObject(context.getString(R.string.display_dimension) + " [dp]", String.format(Locale.ENGLISH, "%.2fx%.2f",
                (160.0 * (float) screenWidth / metrics.densityDpi), (160.0 * (float) screenHeight / metrics.densityDpi)), "dp"));
        list.add(new UIObject(context.getString(R.string.display_diagonal), String.format(Locale.ENGLISH, "%.2f", diagonal), "in"));

        return list;
    }

    static UIObject getXYDpi(Context context, DisplayMetrics metrics) {
        return new UIObject(context.getString(R.string.display_xy_dpi), String.format(Locale.ENGLISH, "%.2fx%.2f", metrics.xdpi, metrics.ydpi));
    }

    static UIObject getOrientation(Context context, Display display) {
        return new UIObject(context.getString(R.string.display_orientation), String.format(Locale.ENGLISH, "%s",
                getOrient(context, display.getRotation())), "°");
    }

    static UIObject getLayoutSize(Context context) {
        return new UIObject(context.getString(R.string.display_size), String.format(Locale.ENGLISH, "%s", getLayoutQualifier(context)));
    }

    static UIObject getType(Context context, Display display) {
        return new UIObject(context.getString(R.string.display_type), String.format(Locale.ENGLISH, "%s", display.getName()));
    }

    static UIObject getDrawType(Context context, DisplayMetrics metrics) {
        return new UIObject(context.getString(R.string.display_draw_size), String.format(Locale.ENGLISH, "%s", densityQualifier(context, metrics.densityDpi)));
    }


    private static String getLayoutQualifier(Context context) {
        Configuration config = context.getResources().getConfiguration();

        switch (config.screenLayout & Configuration.SCREENLAYOUT_SIZE_MASK) {
            case Configuration.SCREENLAYOUT_SIZE_SMALL:
                return context.getString(R.string.display_small);
            case Configuration.SCREENLAYOUT_SIZE_NORMAL:
                return context.getString(R.string.display_normal);
            case Configuration.SCREENLAYOUT_SIZE_LARGE:
                return context.getString(R.string.display_large);
            case Configuration.SCREENLAYOUT_SIZE_XLARGE:
                return context.getString(R.string.display_x_large);
            default:
                return context.getResources().getString(R.string.unknown);
        }
    }

    private static String densityQualifier(Context context, int densityDPI) {

        switch (densityDPI) {
            case DisplayMetrics.DENSITY_LOW:
                return "LDPI";
            case DisplayMetrics.DENSITY_MEDIUM:
                return "MDPI";
            case DisplayMetrics.DENSITY_HIGH:
                return "HDPI";
            case DisplayMetrics.DENSITY_XHIGH:
                return "XHDPI";
            case DisplayMetrics.DENSITY_340:
                return "340DPI XHDPI - XXHDPI";
            case DisplayMetrics.DENSITY_360:
                return "360DPI XHDPI - XXHDPI";
            case DisplayMetrics.DENSITY_400:
                return "400DPI XHDPI - XXHDPI";
            case DisplayMetrics.DENSITY_420:
                return "420DPI XHDPI - XXHDPI";
            case DisplayMetrics.DENSITY_440:
                return "440DPI XHDPI - XXHDPI";
            case DisplayMetrics.DENSITY_XXHIGH:
                return "XXHDPI";
            case DisplayMetrics.DENSITY_560:
                return "440DPI XXHDPI - XXXHDPI";
            case DisplayMetrics.DENSITY_XXXHIGH:
                return "XXXHDPI";
            case DisplayMetrics.DENSITY_TV:
                return "TVDPI";
            case DisplayMetrics.DENSITY_260:
                return "260DPI HDPI - XHDPI";
            case DisplayMetrics.DENSITY_280:
                return "280DPI HDPI - XHDPI";
            case DisplayMetrics.DENSITY_300:
                return "300DPI HDPI - XHDPI";
            default:
                return context.getResources().getString(R.string.unknown);
        }
    }

    private static String getOrient(Context context, int i) {

        switch (i) {
            case 0:
                return "0";
            case 1:
                return "90";
            case 2:
                return "180";
            case 3:
                return "270";
        }
        return context.getResources().getString(R.string.not_available_info);
    }
}

File: app/src/main/java/com/pacmac/devinfo/display/DisplayUtils.kt
package com.pacmac.devinfo.display

import android.content.Context
import android.content.res.Configuration
import android.graphics.Point
import android.hardware.Sensor
import android.hardware.SensorManager
import android.os.Build
import android.util.DisplayMetrics
import android.view.Display
import com.pacmac.devinfo.ListType
import com.pacmac.devinfo.R
import com.pacmac.devinfo.ThreeState
import com.pacmac.devinfo.UIObject
import java.util.Locale

object DisplayUtils {

    const val EXPORT_FILE_NAME = "display_info"

    internal fun getDensity(context: Context, metrics: DisplayMetrics): UIObject =
        UIObject(context.getString(R.string.display_density), String.format(Locale.ENGLISH, "%d", metrics.densityDpi), "dpi")

    internal fun getScaleFactor(context: Context, metrics: DisplayMetrics): UIObject =
        UIObject(context.getString(R.string.display_scale_factor), String.format(Locale.ENGLISH, "%.1f", metrics.density))

    internal fun getRefreshRate(context: Context, display: Display): UIObject =
        UIObject(context.getString(R.string.display_refresh_rate), String.format(Locale.ENGLISH, "%.1f", display.refreshRate), "fps")

    @Suppress("DEPRECATION")
    internal fun getResolution(context: Context, display: Display, metrics: DisplayMetrics): List<UIObject> {
        val size = Point()
        display.getRealSize(size)
        val screenWidth = size.x
        val screenHeight = size.y
        val widthInch = screenWidth.toDouble() / metrics.xdpi.toDouble()
        val heightInch = screenHeight.toDouble() / metrics.ydpi.toDouble()
        val diagonal = Math.sqrt(Math.pow(widthInch, 2.0) + Math.pow(heightInch, 2.0))

        return listOf(
            UIObject(context.getString(R.string.display_resolution), String.format(Locale.ENGLISH, "%dx%d", screenWidth, screenHeight), "px"),
            UIObject(context.getString(R.string.display_dimension), String.format(Locale.ENGLISH, "%.2fx%.2f", widthInch, heightInch), "in"),
            UIObject(context.getString(R.string.display_dimension) + " [dp]", String.format(Locale.ENGLISH, "%.2fx%.2f",
                160.0 * screenWidth.toFloat() / metrics.densityDpi,
                160.0 * screenHeight.toFloat() / metrics.densityDpi), "dp"),
            UIObject(context.getString(R.string.display_diagonal), String.format(Locale.ENGLISH, "%.2f", diagonal), "in"),
        )
    }

    internal fun getXYDpi(context: Context, metrics: DisplayMetrics): UIObject =
        UIObject(context.getString(R.string.display_xy_dpi), String.format(Locale.ENGLISH, "%.2fx%.2f", metrics.xdpi, metrics.ydpi))

    internal fun getOrientation(context: Context, display: Display): UIObject =
        UIObject(context.getString(R.string.display_orientation), getOrient(context, display.rotation), "°")

    internal fun getLayoutSize(context: Context): UIObject =
        UIObject(context.getString(R.string.display_size), getLayoutQualifier(context))

    internal fun getType(context: Context, display: Display): UIObject =
        UIObject(context.getString(R.string.display_type), display.name)

    internal fun getDrawType(context: Context, metrics: DisplayMetrics): UIObject =
        UIObject(context.getString(R.string.display_draw_size), densityQualifier(context, metrics.densityDpi))

    private fun getLayoutQualifier(context: Context): String {
        val config = context.resources.configuration
        return when (config.screenLayout and Configuration.SCREENLAYOUT_SIZE_MASK) {
            Configuration.SCREENLAYOUT_SIZE_SMALL -> context.getString(R.string.display_small)
            Configuration.SCREENLAYOUT_SIZE_NORMAL -> context.getString(R.string.display_normal)
            Configuration.SCREENLAYOUT_SIZE_LARGE -> context.getString(R.string.display_large)
            Configuration.SCREENLAYOUT_SIZE_XLARGE -> context.getString(R.string.display_x_large)
            else -> context.getString(R.string.unknown)
        }
    }

    private fun densityQualifier(context: Context, densityDpi: Int): String = when (densityDpi) {
        DisplayMetrics.DENSITY_LOW -> "LDPI"
        DisplayMetrics.DENSITY_MEDIUM -> "MDPI"
        DisplayMetrics.DENSITY_HIGH -> "HDPI"
        DisplayMetrics.DENSITY_XHIGH -> "XHDPI"
        DisplayMetrics.DENSITY_340 -> "340DPI XHDPI - XXHDPI"
        DisplayMetrics.DENSITY_360 -> "360DPI XHDPI - XXHDPI"
        DisplayMetrics.DENSITY_400 -> "400DPI XHDPI - XXHDPI"
        DisplayMetrics.DENSITY_420 -> "420DPI XHDPI - XXHDPI"
        DisplayMetrics.DENSITY_440 -> "440DPI XHDPI - XXHDPI"
        DisplayMetrics.DENSITY_XXHIGH -> "XXHDPI"
        DisplayMetrics.DENSITY_560 -> "440DPI XXHDPI - XXXHDPI"
        DisplayMetrics.DENSITY_XXXHIGH -> "XXXHDPI"
        DisplayMetrics.DENSITY_TV -> "TVDPI"
        DisplayMetrics.DENSITY_260 -> "260DPI HDPI - XHDPI"
        DisplayMetrics.DENSITY_280 -> "280DPI HDPI - XHDPI"
        DisplayMetrics.DENSITY_300 -> "300DPI HDPI - XHDPI"
        else -> context.getString(R.string.unknown)
    }

    private fun getOrient(context: Context, rotation: Int): String = when (rotation) {
        0 -> "0"
        1 -> "90"
        2 -> "180"
        3 -> "270"
        else -> context.getString(R.string.not_available_info)
    }

    internal fun getHdrCapabilities(context: Context, display: Display): List<UIObject> {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            return listOf(
                UIObject(
                    context.getString(R.string.display_hdr_capabilities),
                    context.getString(R.string.not_available_info)
                )
            )
        }

        val hdrCaps = display.hdrCapabilities
        if (hdrCaps == null || hdrCaps.supportedHdrTypes.isEmpty()) {
            return listOf(
                UIObject(
                    context.getString(R.string.display_hdr_capabilities),
                    context.getString(R.string.display_no_hdr)
                )
            )
        }

        val list = mutableListOf<UIObject>()
        val typeNames = hdrCaps.supportedHdrTypes.map { type ->
            when (type) {
                Display.HdrCapabilities.HDR_TYPE_HDR10 -> "HDR10"
                Display.HdrCapabilities.HDR_TYPE_HLG -> "HLG"
                Display.HdrCapabilities.HDR_TYPE_DOLBY_VISION -> "Dolby Vision"
                else -> {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q && type == Display.HdrCapabilities.HDR_TYPE_HDR10_PLUS) {
                        "HDR10+"
                    } else {
                        "Unknown ($type)"
                    }
                }
            }
        }
        list.add(UIObject(context.getString(R.string.display_hdr_types), typeNames.joinToString(", ")))
        list.add(
            UIObject(
                context.getString(R.string.display_max_luminance),
                String.format(Locale.ENGLISH, "%.1f", hdrCaps.desiredMaxLuminance),
                "nits"
            )
        )
        list.add(
            UIObject(
                context.getString(R.string.display_min_luminance),
                String.format(Locale.ENGLISH, "%.4f", hdrCaps.desiredMinLuminance),
                "nits"
            )
        )
        list.add(
            UIObject(
                context.getString(R.string.display_max_avg_luminance),
                String.format(Locale.ENGLISH, "%.1f", hdrCaps.desiredMaxAverageLuminance),
                "nits"
            )
        )
        return list
    }

    internal fun getWideColorGamut(context: Context, display: Display): UIObject {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            return UIObject(
                context.getString(R.string.display_wide_color_gamut),
                context.getString(R.string.not_available_info)
            )
        }
        val state = if (display.isWideColorGamut) ThreeState.YES else ThreeState.NO
        return UIObject(context.getString(R.string.display_wide_color_gamut), state, ListType.ICON)
    }

    @Suppress("DEPRECATION")
    internal fun getPeakRefreshRate(context: Context, display: Display): List<UIObject> {
        val modes = display.supportedModes
        if (modes.isNullOrEmpty()) {
            return listOf(
                UIObject(
                    context.getString(R.string.display_peak_refresh_rate),
                    context.getString(R.string.not_available_info)
                )
            )
        }
        val peakRate = modes.maxOf { it.refreshRate }
        val modeStrings = modes.map { mode ->
            String.format(Locale.ENGLISH, "%dx%d@%.0ffps", mode.physicalWidth, mode.physicalHeight, mode.refreshRate)
        }
        return listOf(
            UIObject(
                context.getString(R.string.display_peak_refresh_rate),
                String.format(Locale.ENGLISH, "%.1f", peakRate),
                "fps"
            ),
            UIObject(
                context.getString(R.string.display_supported_modes),
                modeStrings.joinToString(", ")
            )
        )
    }

    internal fun getArrSupport(context: Context, display: Display): UIObject {
        if (Build.VERSION.SDK_INT >= 36) {
            return try {
                val hasArr = display.hasArrSupport()
                UIObject(
                    context.getString(R.string.display_arr_support),
                    if (hasArr) ThreeState.YES else ThreeState.NO,
                    ListType.ICON
                )
            } catch (e: Exception) {
                UIObject(context.getString(R.string.display_arr_support), "N/A")
            }
        }
        return UIObject(context.getString(R.string.display_arr_support), "N/A")
    }

    internal fun getSupportedRefreshRates(context: Context, display: Display): UIObject {
        if (Build.VERSION.SDK_INT >= 36) {
            return try {
                val rates = display.supportedRefreshRates
                if (rates != null && rates.isNotEmpty()) {
                    val ratesStr = rates.joinToString(", ") {
                        String.format(Locale.ENGLISH, "%.0f", it)
                    }
                    UIObject(context.getString(R.string.display_supported_refresh_rates), ratesStr, "Hz")
                } else {
                    UIObject(context.getString(R.string.display_supported_refresh_rates), "N/A")
                }
            } catch (e: Exception) {
                UIObject(context.getString(R.string.display_supported_refresh_rates), "N/A")
            }
        }
        return UIObject(context.getString(R.string.display_supported_refresh_rates), "N/A")
    }

    internal fun getHingeAngle(context: Context, sensorManager: SensorManager): UIObject {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
            return UIObject(
                context.getString(R.string.display_hinge_angle),
                context.getString(R.string.not_available_info)
            )
        }
        val hingeSensor = sensorManager.getDefaultSensor(Sensor.TYPE_HINGE_ANGLE)
        return if (hingeSensor != null) {
            UIObject(context.getString(R.string.display_hinge_angle), "Supported")
        } else {
            UIObject(context.getString(R.string.display_hinge_angle), context.getString(R.string.not_available_info))
        }
    }
}
